Update setup instructions with engine source build option (#792)
o3de/o3de.org: content/docs/welcome-guide/setup/setup-from-github.md

toc: true
---

> Getting the source for Open 3D Engine (O3DE) from GitHub is a great way to set up your development environment, so you can easily sync future engine updates and make contributions to the open source project base. Check out the video below for a step by step guide to set up O3DE from GitHub, or follow the tutorial on this page.

Reviewer: ```suggestion
Getting the source for Open 3D Engine (O3DE) from GitHub is a great way to set up your development environment. This method lets you easily sync future engine updates and make contributions to the open source project base. For instructions on setting up O3DE from GitHub, check out the following video or the tutorial on this page.
```
* For the first sentence:
    * It's a long one. I recommend splitting it in two for flow, readability.
    * Be cautious around using "a great way" and "easily". These can sound a bit like Marketing-speak.
* Re-ordered the last sentence to for flow, focus.

> The instructions here and in the video will guide you through the following steps:

Reviewer: ```suggestion
The video and written instructions guide you through the following steps:
```

    upstream        {{< links/o3de-source >}}.git (push)
    ```

> 1. Update the LFS URL to include your fork. This will enable you to push changes to large files. Open the .lfsconfig file at the root of the repository for complete instructions and the **DISTRIBUTION** to use.

Author: ```suggestion
1. (Optional) Update the LFS URL to include your fork. This will enable you to push changes to files in the LFS. If you do not expect to make changes to these files, you do not need to do this step. Open the .lfsconfig file at the root of the repository for complete instructions and the **DISTRIBUTION** to use.
```

> 1. (Optional) Update the LFS URL to include your fork. This will enable you to push changes to files in the LFS. If you do not expect to make changes to these files, you do not need to do this step. Open the .lfsconfig file at the root of the repository for complete instructions and the **DISTRIBUTION** to use.

Reviewer: ```suggestion
1. (Optional) If you don't expect to make changes to these files, ignore this step. Otherwise, update the LFS URL to include your fork. This lets you push changes to files in the LFS. For complete instructions and the **DISTRIBUTION** to use, open the `.lfsconfig` file at the root of the repository.
```

    git config --unset lfs.url 
    ```

    {{< note >}}

> If you did not clone from a fork, you do not need to do this step.

Author: ```suggestion
If you did not clone from a fork, or will not push changes to files in the LFS, you do not need to do this step.
```

>     {{< note >}}
> If you did not clone from a fork, or will not push changes to files in the LFS, you do not need to do this step.
>     {{< /note >}}
> 

Author: ```suggestion
```

For more information and examples of common contributor workflows, refer to [O3DE Code Contribution GitHub Workflow](/docs/contributing/to-code/git-workflow) in the Contributor Guide.

## Build the engine

> Now that you have a local copy of the O3DE source, you can build the engine, including key tools such as the O3DE **Asset Processor**, **Editor**, and **Project Manager**.

Reviewer: ```suggestion
Now that you have a local copy of the O3DE source, you can build the engine, including key tools such as **Asset Processor**, **O3DE Editor**, and **Project Manager**.
```

Choose one of the following build types based on the primary focus of your development work, then follow the instructions in the corresponding tab:

1. **Source Engine** - Choose this build type if you plan to make frequent changes to the engine source code.

> 1. **Pre-Built SDK Engine** - Choose this build type if you are primarily interested in project development and make infrequent or no changes to the engine source.

Reviewer: ```suggestion
1. **Pre-built SDK Engine** - Choose this build type if you're primarily interested in project development and you plan to make only infrequent changes (or no changes) to the engine source.
```

{{< tabs name="Engine build instructions" >}}

> {{% tab name="Source Engine" %}}

Reviewer: ```suggestion
{{% tab name="Source engine" %}}
```

> 1. Create a package directory in a writeable location. The directory `C:\o3de-packages` is used in the examples that follow.

Reviewer: ```suggestion
1. Create a package directory in a writeable location. The following examples use the directory `C:\o3de-packages`.
```

    ```cmd
    mkdir C:\o3de-packages
    ```

>     This directory will be used by the O3DE package downloader to retrieve external libraries needed for the engine.

Reviewer: ```suggestion
The O3DE package downloader uses this directory to retrieve external libraries needed for the engine.
```

> 1. Get the Python runtime. The Python runtime is not included in the GitHub repo. It is required by the `o3de` script, which you will use to perform common command line functions. This script requires **CMake** to be installed and accessible on your device's path. If you have not installed CMake, or get an error that CMake cannot be found when running the script, refer to the [O3DE System Requirements](../requirements) page for installation instructions.

Reviewer: ```suggestion
1. Get the Python runtime, which isn't included in the GitHub repo. The `o3de` script requires this runtime. You'll use this script to run common command line functions. This script also requires **CMake** to be installed and accessible on your device's path. If you haven't installed CMake, or you get an error that CMake cannot be found when running the script, refer to the [O3DE System Requirements](../requirements) page for installation instructions.
```

>     Open a command prompt to the directory where you set up O3DE and run the `get_python` script.

Reviewer: ```suggestion
Open a command prompt and change to the directory where you set up O3DE, then run the `get_python` script.
```

    ```cmd
    python\get_python.bat
    ```

1. Use CMake to create the Visual Studio project for the engine. Supply the build directory, the Visual Studio generator, the path to the packages directory that you created, and any other project options. Paths can be absolute or relative. Alternatively, you can use the CMake GUI to complete this step.

    ```cmd
    cmake -B build/windows_vs2019 -G "Visual Studio 16" -DLY_3RDPARTY_PATH=C:\o3de-packages -DLY_UNITY_BUILD=ON
    ```

>     There are several noteworthy custom definitions (`-D`) specified in the preceding command. All are optional but recommended in this example.

Reviewer: ```suggestion
The preceding command specifies several noteworthy custom definitions (`-D`). All are optional but recommended in this example.
```

    * `LY_3RDPARTY_PATH` : The path to the downloadable package directory, also known as the "third-party path". Do not use trailing slashes when specifying the path to the packages directory.

>     * `LY_UNITY_BUILD` : Unity builds are recommended in many cases for improved build performance. If you encounter a build error, disable unity builds to help debug the problem.

Reviewer: ```suggestion
    * `LY_UNITY_BUILD` : We recommend Unity builds in many cases for improved build performance. If you encounter a build error, disable Unity builds to help debug the problem.
```
Changed to "We recommend" for active voice. If that's not accurate (and "are recommended" is just generally true), decline that change.

> 1. (Optional) Use CMake to build the source engine. This step is optional because in the "source engine" build model, the engine is built inside of every _project_. To avoid building the engine twice if you plan on working with projects, you might want to wait until you learn how to create and build a project in a later section on creating projects. However, if you want to proceed and build the engine without a project, use the following CMake build command.

Reviewer: ```suggestion
1. (Optional) Use CMake to build the source engine. This step is optional because in the "source engine" build model, the engine is built inside of every project. If you plan on working with projects, to avoid building the engine twice, consider waiting until you learn how to create and build a project, which our documentation on creating projects covers. The following command builds the engine without a project.
```
* I don't think that the italics are necessary for emphasis here. If you prefer to keep them, consider whether the whole phrase "every project" should be italicized instead.
* Other suggested edits for flow and to reduce a bit of redundancy.

>     The `profile` build configuration is shown in this example.

Reviewer: ```suggestion
The following example shows the `profile` build configuration.
```

    ```cmd
    cmake --build build/windows_vs2019 --target Editor --config profile -- /m
    ```

>     The `/m` is a recommended build tool optimization, which tells the Microsoft compiler (MSVC) to use multiple threads during compilation to speed up build times.

Reviewer: ```suggestion
The `/m` is a recommended build tool optimization. It tells the Microsoft compiler (MSVC) to use multiple threads during compilation to speed up build times.
```

Author: The sentence here is relevant to the step above it. IMHO, the indentation here helps connect the sentence to that step.

Reviewer: Good call. Ignore the indentation change here. The rest of the change is consistent with the same sentence earlier in this PR.

>     The engine will take a while to build. If you've used all the example commands in these steps, when the build is complete, the engine tools and other binaries can be found in `C:\o3de\build\windows_vs2019\bin\profile`.

Reviewer: ```suggestion
The engine takes a while to build. If you've used all the example commands in these steps, when the build is complete, you can find the engine tools and other binaries in `C:\o3de\build\windows_vs2019\bin\profile`.
```

{{% /tab %}}

> {{% tab name="Pre-Built SDK Engine" %}}

Reviewer: ```suggestion
{{% tab name="Pre-built SDK engine" %}}
```

    cmake --build build/windows_vs2019 --target INSTALL --config profile -- /m
    ```

    The `/m` is a recommended build tool optimization, which tells the Microsoft compiler (MSVC) to use multiple threads during compilation to speed up build times.

> The engine will take a while to build. If you've used all the example commands in these steps, when the build is complete, the engine tools and other binaries can be found in `C:\o3de-install\bin\Windows\profile`.

Reviewer: ```suggestion
The engine takes a while to build. If you've used all the example commands in these steps, when the build is complete, you can find the engine tools and other binaries in `C:\o3de-install\bin\Windows\profile`.
```
Forgot to apply this edit to the parallel outro sentence on this tab. (Same edit as the other tab.)

Author: Whoops, I should have checked for that too.

{{% /tab %}}
{{< /tabs >}}

## Register the engine

Registering the O3DE engine enables O3DE projects to find the engine, even when they exist in different locations on your computer. The registration process creates (or updates) the **O3DE manifest** in your user directory.

Choose the tab that corresponds to the engine build type you chose in the preceding set of instructions.

{{< tabs name="Engine registration instructions" >}}

> {{% tab name="Source Engine" %}}

Reviewer: ```suggestion
{{% tab name="Source engine" %}}
```

1. Open a command window if you don't already have one open. Change your current directory to the source engine directory.

    ```cmd
    cd C:\o3de
    ```

1. Use the `o3de` script to register the engine.

    ```cmd
    scripts\o3de.bat register --this-engine
    ```

    The O3DE manifest file is `<USER_DIRECTORY>/.o3de/o3de_manifest.json`. The paths to all the registered engines, projects, and more are recorded in this file.

{{% /tab %}}

> {{% tab name="Pre-Built SDK Engine" %}}

Reviewer: ```suggestion
{{% tab name="Pre-built SDK engine" %}}
```

    The O3DE manifest file is `<USER_DIRECTORY>/.o3de/o3de_manifest.json`. The paths to all the registered engines, projects, and more are recorded in this file.

{{% /tab %}}
{{< /tabs >}}

> You are now ready to create a project! For an introduction to project configuration, refer to [Project Creation with Open 3D Engine](/docs/welcome-guide/create).

Reviewer: ```suggestion
You're now ready to create a project. For an introduction to project configuration, refer to [Project Creation with Open 3D Engine](/docs/welcome-guide/create).
```

## Side-by-side engines
 
 In the preceding sections, you learned how to set up and configure the O3DE engine as either a source engine or a pre-built SDK engine. You also have the option of registering both engines in a side-by-side configuration. This is one way to isolate your engine source code development from your project development.

>  For example, if the engine source is in `C:\o3de` and the pre-built SDK engine is in `C:\o3de-install`, you can give each engine its own engine name so that you can register both in the O3DE manifest using the `o3de` script's `register` command. To test your engine modifications, you can build a test project using the source engine project creation instructions, which builds the engine in the project directory. When you're ready to update the SDK engine used by your production project, you can build a new version of the SDK engine from `C:\o3de` using the `INSTALL` target. This will update the binaries in `C:\o3de-install`.

Reviewer: ```suggestion
 For example, if the engine source is in `C:\o3de` and the pre-built SDK engine is in `C:\o3de-install`, you can give each engine its own engine name so that you can register both in the O3DE manifest using the `o3de` script's `register` command. To test your engine modifications, you can build a test project using the source engine project creation instructions, which builds the engine in the project directory. When you're ready to update the SDK engine that your production project uses, you can build a new version of the SDK engine from `C:\o3de` using the `INSTALL` target. This updates the binaries in `C:\o3de-install`.
```

> To update the name of an engine, open `engine.json` in the engine root directory and change the `engine_name` field. Then run `scripts\o3de.bat register --this-engine` again from that directory. New projects created from this engine will be configured to use the new engine name. To update the configured engine for an existing project, open `project.json` in the project root directory and change the `engine` field to use the new engine name.

Reviewer: ```suggestion
To update the name of an engine, open `engine.json` in the engine root directory and change the `engine_name` field. Then run `scripts\o3de.bat register --this-engine` again from that directory. New projects that you create from this engine will use the new engine name. To update the configured engine for an existing project, open `project.json` in the project root directory and change the `engine` field to use the new engine name.
```
Not sure if "will be configured to use" is a key detail in the third sentence. Add it back if that's important.

Author: I think that key detail is still implied by your active-voice rewrite.
